sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,493 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Search.Documents.Batching
+{
+    /// <summary>
+    /// Implement a Producer/Consumer pattern where we (potentially) have
+    /// multiple document producers and a single consumer that submits those
+    /// events to a service.  This class is mostly focused on synchronizing
+    /// behavior between document producers and relies on abstract methods to
+    /// implement service specific semantics.
+    /// </summary>
+    /// <typeparam name="T">The type of document being published.</typeparam>
+    internal abstract partial class Publisher<T> : IDisposable, IAsyncDisposable
+    {
+        /// <summary>
+        /// Flag indicating whether the publisher has been disposed.
+        /// </summary>
+        private volatile int _disposed = 0;
+
+        /// <summary>
+        /// Channel used to communicate between the sender and publisher.
+        /// </summary>
+        private Channel<Message> _channel = Channel.CreateUnbounded<Message>(
+            new UnboundedChannelOptions { SingleReader = true });
+
+        /// <summary>
+        /// "Blocking" semaphore used to wait for flushing to complete.
+        /// </summary>
+        private ProducerBarrierSlim _flushBarrier = new ProducerBarrierSlim();
```

Comment:
Can we use the TaskCompletionSource for this? The reset would be replacing the source with an incomplete one and release signaling the source.

Reply:
After discussing, I think I can make that work.

Reply:
Changed, but the problem now is we're no longer waiting on the user's cancellation token.  Is there a way to have the user's token respected when awaiting the completion source?  Everything I'm thinking of feels really hacky.

Reply:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Shared/TaskExtensions.cs#L18